Email editor- register_block_template [MAILPOET-6356]

costasovo opened the pull request:
## Description

This PR is refactoring the templates to the new `register_block_template` function added in WP 6.7

## Code review notes

Please review commit by commit.
I had to update WordPress stubs for PHPStan because the new function was in the old stubs.
You may notice this change and a similar one with getPostTypes:
```php
- $this->wp->getTaxonomies([], 'object'),
+ $this->wp->getTaxonomies([], 'objects'),
```
It is safe because,[ internally, anything else than `name` or `names` is processed as `objects`.](https://github.com/WordPress/wordpress-develop/blob/e58bef6fdc10bbbbba02e58a34dbca022e956543/src/wp-includes/taxonomy.php#L285)

## QA notes

The templates should work the same as before.
You can test:

- creating a new email
- swapping template 
- editing template

## Linked PRs

https://github.com/mailpoet/mailpoet-premium/pull/930

## Linked tickets

[MAILPOET-6356]

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-6356]: https://mailpoet.atlassian.net/browse/MAILPOET-6356?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Preview

[Preview in WordPress Playground](https://account.mailpoet.com/playground/new/branch:email-editor/register-templates)

_The latest successful build from `email-editor/register-templates` will be used. If none is available, the link won't work._

costasovo committed:
Switch template registration to use register_block_template

[MAILPOET-6356]

costasovo committed:
Remove code that is no longer necessary for the templates to work

[MAILPOET-6356]

costasovo committed:
Fix a typo in a method name in Email_Editor

[MAILPOET-6356]

costasovo committed:
Add post_types field to the template REST API response

We want to use post_type for filtering email templates.
There is a PR that adds this property https://github.com/WordPress/wordpress-develop/pull/7530
but it is not merged.
This commit added the property and also future compatibility check
so that we don't attempt to add it after the PR is merged.
[MAILPOET-6356]

costasovo committed:
Move the template_id construction logic to Templates class

[MAILPOET-6356]

costasovo committed:
Add workaround for saving post-template association for non-block themes

[MAILPOET-6356]

costasovo committed:
Add action for registering email templates

This commit introduces a new action to trigger registering email templates.
We need a custom action for that because the email editor engine controls
when we register templates and when we don't. Currently, we still skip
registering templates when we detect the site editor page.
[MAILPOET-6356]

costasovo committed:
Move simple-light template from the engine/package to the free plugin

The additional templates are registered by integrators.
This way, we will be able to use assets like images from MailPoet's CDN
and also various services from the plugins (e.g. CdnAssetsController)
[MAILPOET-6356]

costasovo committed:
Use registered email post types for templates instead of hardcoded mailpoet_email

The general email template is registered and works with all post types that are
marked as email post types during editor initialization.
[MAILPOET-6356]

costasovo committed:
Update szepeviktor/phpstan-wordpress for PHPStan

We need to udpate WordPress Stubs to recognize new register_block_template
function.
[MAILPOET-6356]

costasovo committed:
Fix PHPStan errors after updating WordPress Stubs

[MAILPOET-6356]

lysyjan reviewed (dismissed):
Great job @costasovo! 
You added a lot of valuable comments. The PR looks good to me.

costasovo committed:
Fix saving template associations with emails

The original workaround was not working for templates that have edits saved in the DB.
The problem was that when the template is saved in the DB the get_block_templates()
returns WP_Block_Template object with post_types as null.
It seems this association is not saved to DB in the core.
[MAILPOET-6334]

costasovo committed:
Improve the stability of the EmailTemplatesCest

[MAILPOET-6356]
